### PR TITLE
`sqrt` implementation

### PR DESCRIPTION
**Pull request by squ1dd13**
I recently found myself needing `sqrt` while using this crate, and after reading #60 I decided to have a go at implementing it myself.

The implementation is quite long because it's the result of quite a lot of benchmarking to try and get decent performance, so I've added a new `approx` module for it. The API isn't very polished yet but the actual algorithms work fairly well and are able to produce very accurate rational approximations for square roots in relatively little time. I thought I'd just get an extra opinion on it before I spend time perfecting the API, which is why I've only submitted a draft PR.

Let me know if you think this is worth finishing off. Cheers :)

### CODE REVIEW

File: src/fraction/approx.rs [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+}
+
+impl<T: Clone + Integer + ToBigInt + GenericInteger> GenericFraction<T> {
+    /// Returns an unsimplified rational approximation of the square root of `self`.
+    ///
+    /// If you need the result to be simplified, use `sqrt_with_accuracy` instead.
+    ///
+    /// The square of the resulting value is guaranteed to be equal to `self` to the number of
+    /// decimal places specified by `accuracy`.
+    pub fn sqrt_with_accuracy_raw(&self, accuracy: &SqrtAccuracy) -> SqrtApprox {
```
Comment (dnsl48) [inline src/fraction/approx.rs:218]: Perhaps, accuracy could accept `Option<&GenericInteger>`?
It could accept a pre-existing BigUint/BigInt as well as a new stack (or lazy_static) value? `None` would be provided by `Default` implementation for `SqrtAccuracy`?

**Review by dnsl48 — COMMENTED** [review]
Hi @squ1dd13,
Thank you for the PR, it looks great!
Overall it looks awesome. I have left some _opinionated_ feedback on the patch, please feel free to push back on any of the suggestions. Also, happy to keep discussing if you have more thoughts on any of subjects.
Looking forward for the polished version! :)

**Comment by squ1dd13** [comment]
Thanks for all the feedback, I'm glad you're interested :D

Since there are several points to address regarding `SqrtAccuracy`, I'll just reply here rather to the individual comments. The main rationale behind allowing reuse of `SqrtAccuracy` objects was that since the `multiplier` is heap-allocated, there might be some cases where performing several `sqrt_*` calls would be able to avoid heap-allocating a new `multiplier` on each call. After having another look it seems like since `multiplier` gets cloned anyway, this (premature) optimisation is unnecessary. I could benchmark it, but I suspect it's probably worth very little compared to other optimisations that might be possible within the main algorithm. If nothing else, this would simplify the API quite a lot.

TL;DR I'm all for getting rid of `SqrtAccuracy` entirely.

File: src/fraction/mod.rs [hunk 2]
```diff
@@ -1,6 +1,7 @@
 pub use self::generic_fraction::GenericFraction;
 pub use self::sign::Sign;
 
+pub mod approx;
```
[the remark this answers is not in the document]
Comment (squ1dd13) [reply src/fraction/mod.rs:4]: For sure. Do you think it would make sense to just keep `approx` behind `with-bigint`, or should there be a separate feature entirely for `approx`? The latter seems reasonable to me on the grounds that this crate is primarily focused around lossless maths (ergo many users probably won't care about `approx`) but then again it's an extra feature flag.

File: src/fraction/approx.rs [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+}
+
+/// An approximation of a square root.
+pub enum SqrtApprox {
+    /// A rational (i.e. fractional) approximation.
+    ///
+    /// Depending on the accuracy used, these numbers can be *very* large to store (>100 KB with
+    /// excessive accuracy), so cloning is likely to be expensive. This size also means that
+    /// computations can take a long time.
+    Rational(BigRational),
```
[the remark this answers is not in the document]
Comment (squ1dd13) [reply src/fraction/approx.rs]: I had done this originally, but I switched to `BigRational` because it's easier to work with (it has `from_float`, for example). I've just switched it all back and it wasn't as difficult as I thought, so thanks for the suggestion.

File: src/dynaint.rs [hunk 4]
```diff
@@ -886,6 +887,19 @@ where
     dyna_impl!(impl_fn_refmath_tuple2self; div_rem);
 }
 
+impl<T, G> ToBigInt for DynaInt<T, G>
```
[the remark this answers is not in the document]
Comment (squ1dd13) [reply src/dynaint.rs:891]: Thanks. I've added this on `ToBigInt` and a new `ToBigUint` as well as for the whole `approx` module (for now; see [my comment below](https://github.com/dnsl48/fraction/pull/84#discussion_r1314323501)).

[the remark this answers is not in the document]
Comment (squ1dd13) [reply src/dynaint.rs:891]: I've now changed things so that the `sqrt_*` methods directly return `GenericFraction<BigUint>` (rather than `DynaFraction`, which was what it was before). This is because
1. Any reasonable approximation will have a numerator and denominator so large that it would not be possible to store them even in `u128`, making `DynaFraction` essentially equivalent to `GenericFraction<BigUint>` anyway;
2. The algorithm itself works with `BigUint`s, so returning a fraction made of `BigUint`s means there's one less processing step before output; and
3. If the user wants a `DynaFraction`, then they can do that for themselves.

However, I've left the implementations of `ToBigInt` and `ToBigUint` for `DynaInt` because they allow `sqrt` for `DynaFraction`.

**Comment by squ1dd13** [comment]
For fractions I think it's probably essential to force the user to recognise that they are only working with an approximation, hence the requirement for them to supply some sort of value specifying the accuracy they want the result to have. However, would it make sense for `GenericDecimal` to also provide a `sqrt` function, but perhaps one which uses the precision type to automatically pick an accuracy level?

This would be somewhat at odds with the current documentation, which states

> Calculations do not use precision

although the other part of this,

> but comparison does

is actually exactly what the implementation of `sqrt` is doing already. We use a truncated comparison between the square of our approximation of the root and the original value we were given to determine when to stop iterating:
```rs
// Stop and yield the current guess if it's close enough to the true value.
if squared_and_truncated == truncated_target {
    break SqrtApprox::Rational(current_approx);
}
```

File: src/fraction/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (dnsl48) [reply src/fraction/mod.rs:5]: I feel like a separate feature would be fine, as long as it depends on `with-bigint` and that dependency is documented. The vice-versa, people who only need a bigint and not approx would be able to build the lib without dead code and get a smaller size.

File: src/dynaint.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (dnsl48) [reply src/dynaint.rs:891]: that sounds good, thank you!

**Comment by dnsl48** [comment]
@squ1dd13, I can't reply in threads to the posts in the general comment section, so I'll quote you in this reply to be more specific about things.

> Since there are several points to address regarding SqrtAccuracy, I'll just reply here rather to the individual comments. The main rationale behind allowing reuse of SqrtAccuracy objects was that since the multiplier is heap-allocated, there might be some cases where performing several sqrt_* calls would be able to avoid heap-allocating a new multiplier on each call. After having another look it seems like since multiplier gets cloned anyway, this (premature) optimisation is unnecessary.

That makes a lot of sense. I agree that it sounds reasonable to pass a link to a pre-allocated instance and do not destroy it for future reuse. Also, I guess it could make sense to have a _default_ lazy-static pre-allocated value for it as a fallback.

> but I suspect it's probably worth very little compared to other optimisations that might be possible within the main algorithm. If nothing else, this would simplify the API quite a lot.

Perhaps, all possible optimisations have value regardless of their relative size. If they obscure API too much, then we may need to think about adding abstractions to simplify that (good thing Rust supports zero-cost abstractions really well :)

> TL;DR I'm all for getting rid of SqrtAccuracy entirely.

I feel like that could be just a ref to GenericInteger in there.

> However, would it make sense for GenericDecimal to also provide a sqrt function, but perhaps one which uses the precision type to automatically pick an accuracy level?

Decimal is implied to be a thin wrapper around Fraction, which contains extra info about how we want the fraction to be _represented_ in a decimal format.
E.g.
 - `(1/3, precision=3) == 0.333`
 - `(1/3, precision=4) == 0.3333`

The comparison takes precision into account precision to follow the "string" way of comparing numbers (e.g. 1/3 == 1/3, but 0.333 != 0.3333 while both of those might be represented as the same fraction internally).

Given that, sqrt itself is not producing a _formatted representation of the decimal itself_, but rather a _calculated output`, maybe `precision` is not always be the best value for `accuracy` for the sqrt. However, it is hard to tell without a real example.
I would say, up to you if you feel that is a good default or not.

File: src/fraction/approx.rs [hunk 5]
```diff
@@ -4,51 +4,181 @@
...
+        }
+    }
+
+    /// Returns a [`BigFraction`] which is equal to `fraction` according to this `Accuracy` by
+    /// "chopping off" any irrelevant digits.
+    ///
+    /// The result will be equal to `(fraction * self.multiplier()).floor() / self.multiplier()`.
+    ///
+    /// This method propagates infinity and NaN values.
+    pub fn chop<T>(&self, fraction: &GenericFraction<T>) -> BigFraction
```
Comment (squ1dd13) [inline src/fraction/approx.rs:101]: I've exposed a minimal public API for `Accuracy` so that it might be useful for users writing their own approximation code (or indeed anything else).

File: Cargo.toml [hunk 6]
```diff
@@ -52,6 +52,7 @@ with-serde-support = ["serde", "serde_derive", "num/serde"]
 [dev-dependencies]
 criterion = "0.4"
 rand = "0.8.5"
+lazy_static = "1"
```
Comment (squ1dd13) [inline Cargo.toml:55]: This absolutely isn't a hard requirement for the `approx` tests, but it's useful. Happy to remove it.

File: src/fraction/approx.rs [hunk 7]
```diff
@@ -441,38 +443,92 @@ impl<T: Clone + Integer + ToBigUint + ToBigInt + GenericInteger> GenericFraction
...
+                println!("sqrt({value}) to {accuracy} d.p., simplified");
+                test_sqrt_of(&value, true, accuracy);
+            }
+        }
+    }
 
-        let s = format!("{sqrt:.100}");
-        assert_eq!(s, "1.0943425455728974838600903859180783076888376493725431295813967125637781050743787384965051763360943812");
+    #[test]
+    fn test_raw() {
```
Comment (squ1dd13) [inline src/fraction/approx.rs:533]: I've split the tests into simplified and raw versions. Unfortunately, simplifying takes a very long time: `test_raw` completes in just under 12.5 seconds on my machine; `test_simplified` takes about 39 seconds. I can imagine that the former might be tolerable, but the latter seems excessive.

Do we need to test the simplified results, considering that it's just the same as `test_raw` but applying `num`'s simplifying routine on each value? My instinct is to test everything, but here I think maybe it would be fine not to?

**Comment by dnsl48** [comment]
Hi @squ1dd13,

Thank you for the detailed explanation. Yes, agree, having `SqrtAccuracy` type makes sense. Also agree that we can have `::default` instead of `None`. My thinking about the `Option` was to use a _lazy_static_ value in case of `None` instead of allocating a new value for each call. Perhaps, we could achieve something similar within the SqrtAccuracy internal logic. Not sure if that's worth pursuing though. I wouldn't mind accepting the patch without that optimisation.

I also feel that might be important to capture the details from your above post in the documentation for the SqrtAccuracy structure and the sqrt function.

File: src/fraction/approx.rs [hunk 8]
```diff
@@ -0,0 +1,541 @@
...
+    ///
+    /// # Panics
+    /// This method will panic if `self` is negative.
+    pub fn sqrt_with_accuracy_raw(&self, accuracy: impl Borrow<Accuracy>) -> SqrtApprox {
+        let accuracy = accuracy.borrow();
+
+        let SqrtSetup {
+            estimate: initial_estimate,
+            value_as_ratio,
+        } = SqrtSetup::for_value(self);
```
Comment (dnsl48) [inline src/fraction/approx.rs:359]: I wouldn't mind a private method to panic, but for a public one it feels like an overkill?
Perhaps we could return an `Option<>` here to play it safe?
Or maybe even fall back to `NaN` if we want to spare some extra bits?

Otherwise, it might be worth making the method _unsafe_ and let it panic, but then create another one which would produce an Option and be _safe_?

File: src/fraction/approx.rs (hunk 7 above)
Comment (dnsl48) [reply src/fraction/approx.rs:533]: Agree, if we rely on the external `simplify` functionality, we should be fine to assume it works as expected and only test the raw versions.
On the other hand, I don't think 39 seconds is a totally unaccepted. I think we should be fine to take it as long as it's below 1 minute to run (for one module).

**Review by dnsl48 — CHANGES_REQUESTED** [review]
That is a great patch, thank you for submitting!
I would be keen to have safer handling for the negatives by default. At least the default API method shouldn't panic so we could handle unexpected data safely, and maybe then provide unsafe but faster method for those who know what they are doing.
Also, would be nice to cover all possible values in the tests:
 - NaN
 - -Inf
 - -2
 - 0
 - 2
 - +Inf

**Comment by squ1dd13** [comment]
> My thinking about the Option was to use a lazy_static value in case of None instead of allocating a new value for each call. Perhaps, we could achieve something similar within the SqrtAccuracy internal logic.

Is this not covered by the behaviour of the `Accuracy` type? When `lazy_static` is enabled, the user can make use of the `Dp20`, `Dp100` and `Dp500` variants, which do not store a `BigUint` and instead use pre-allocated `lazy_static` `BigUint` values:

```rs
/// Returns a reference to the multiplier used by `self` to chop off irrelevant digits.
pub fn multiplier(&self) -> &BigUint {
    #[cfg(feature = "lazy_static")]
    {
        lazy_static! {
            static ref DP20_MUL: BigUint = BigUint::from(10_u8).pow(20_u32);
            static ref DP100_MUL: BigUint = BigUint::from(10_u8).pow(100_u32);
            static ref DP500_MUL: BigUint = BigUint::from(10_u8).pow(500_u32);
        };

        return match self {
            Accuracy::Dp20 => &DP20_MUL,
            Accuracy::Dp100 => &DP100_MUL,
            Accuracy::Dp500 => &DP500_MUL,
            Accuracy::Custom { multiplier } => multiplier,
        };
    }

    // -snip-

    let Accuracy::Custom { multiplier } = self else {
        // -snip-
    };

    multiplier
}
```

The current `Default` implementation for `Accuracy` is just `Dp100`, so if the user passes `Accuracy::default()`, there will be no allocation. If `lazy_static` is disabled, only the `Accuracy::Custom` variant is available*. (This is better than leaving the `DpX` variants enabled and adapting `multiplier` to allocate a value for them on-the-fly, because `::Custom.multiplier()` doesn't need to allocate, but the `DpX` variants don't store a `BigUint` inline so without `lazy_static` they will always have to allocate a new one when it is needed.)

*This means `Accuracy::default` needs `lazy_static` at the moment. It might be more ergonomic to just change `default` so that it allocates a new `Accuracy::Custom` when `lazy_static` is off, but I'm not sure how comfortable I feel with hiding a heap allocation behind a `default` call.

> I also feel that might be important to capture the details from your above post in the documentation for the SqrtAccuracy structure and the sqrt function.

Absolutely. I'll write up full documentation for each of the `sqrt*` methods and include this. As `SqrtAccuracy` has become the more general `Accuracy`, I think the specific way in which the `Accuracy` value passed is used should be up to the individual approximations implemented in the module, and not the documentation of the `Accuracy` type itself. (At the moment the only approximation is `sqrt`, but Newton's method could be used for fractional `pow` amongst other things. The reason I've only implemented `sqrt` so far is because it's ubiquitous enough that it makes sense to implement it by itself to take advantage of the specific optimisations that it allows. That is, `halve_in_place_pos_rational` and any other specialised stuff.)

File: src/fraction/approx.rs (hunk 8 above)
Comment (squ1dd13) [reply src/fraction/approx.rs:359]: Good idea.

Since we're working with high-level types here which explicitly encode the sign, I can't see a reasonable way to implement an `_unchecked` (for example) version that would use unsafe code. There's not really any scope for us to invoke undefined behaviour here.

Perhaps we could have something safe like `sqrt_abs`, which would only calculate the square root of the magnitude value of `self`? Then we can just destructure `self` to obtain its magnitude and ignore the sign (since the two are separate). This would be nice because it's not an inherently unsafe operation, and it seems fairly likely that there will be many instances where the programmer knows that `self` is not negative – it would be annoying to have to use unsafe code to avoid the check in situations like that.

File: src/fraction/approx.rs (hunk 7 above)
Comment (squ1dd13) [reply src/fraction/approx.rs:533]: Great, thanks. I'll get rid of `test_simplified` and perhaps extend `test_raw` to be more comprehensive, given that we've got a bit more time to spare.

File: src/fraction/approx.rs (hunk 8 above)
Comment (dnsl48) [reply src/fraction/approx.rs]: > Perhaps we could have something safe like sqrt_abs

sounds like a nice workaround. I like it!

> There's not really any scope for us to invoke undefined behaviour here.

Agree. I guess, my main concern is not UB, but panic which might be caused by user input. Panic is not something that the library users would be able to _catch_ outside of the library, so they would have to manually check the sign everywhere to not let the application crash.

**Comment by dnsl48** [comment]
> Is this not covered by the behaviour of the Accuracy type? When lazy_static is enabled, the user can make use of the Dp20, Dp100 and Dp500 variants,

Yes, that looks good. I was writing that comment before reviewing the latest changes to the PR, sorry for the confusion.

> Accuracy::default needs lazy_static at the moment

Totally happy with that. There are no strong arguments for not having `lazy_static` as a dependency. We also have it now required by `with-bigint` feature anyway. And bigint is a hard requirement for the approx module. Thus, lazy_static is an indirect requirement of approx anyway.

**Comment by squ1dd13** [comment]
> There are no strong arguments for not having `lazy_static` as a dependency. We also have it now required by `with-bigint` feature anyway. And bigint is a hard requirement for the approx module. Thus, lazy_static is an indirect requirement of approx anyway.

I keep forgetting this :P I'll leave the explicit requirement for `lazy_static` on those variants in case we end up changing things to make `with-bigint` optional for `approx` in the future.

I think the PR is nearly done. I'll write some more tests so that we have all the special cases covered, and then write up the documentation for all of the methods. Please do say if there's anything that the docs leave unclear – I'm used to the inner workings of the module now so there's a chance I'll miss out bits that I forget might not be obvious.

Not directly related to this PR as such but still potentially important is that there are a couple of issues I noticed with `GenericFraction<T>` allowing signed `T`. I think the best summary of this is given by the [implementation of `sqrt_with_accuracy_raw` as of yesterday](https://github.com/squ1dd13/fraction/blob/b1fd030cfac021b2efd1038fddda7fe379a77bdc/src/fraction/approx.rs#L482), but in short: stuff like `GenericFraction::<T>::signum`, `::is_sign_positive`, `::is_sign_negative`, etc. cannot be trusted for signed `T`.

**Comment by squ1dd13** [comment]
That's everything I can think of for now. I've made `sqrt` its own module under `approx` so that it has module-level documentation where all of the concepts (the whole accuracy thing, `raw`, `with_accuracy`, `abs`) can be discussed in one place. This makes documenting the individual methods much easier.

I plan to open at least one more PR with other approximations, so it would make sense for those to then be their own modules within `approx`. `Accuracy` is a child of `approx`, so it would be available to all of the approximation modules.

**Comment by dnsl48** [comment]
> Not directly related to this PR as such but still potentially important is that there are a couple of issues I noticed with GenericFraction<T> allowing signed T. I think the best summary of this is given by the [implementation of sqrt_with_accuracy_raw as of yesterday](https://github.com/squ1dd13/fraction/blob/b1fd030cfac021b2efd1038fddda7fe379a77bdc/src/fraction/approx.rs#L482), but in short: stuff like GenericFraction::<T>::signum, ::is_sign_positive, ::is_sign_negative, etc. cannot be trusted for signed T.

This is true, the negative numbers within the Ratio data (either num or denum) would lead to destructive consequences in many scenarios. Partially, this is neglected because the GenericFraction implementation keeps the sign out as a separate attribute: `enum ... Rational(Sign, Ratio<T>)`. Because of that, keeping signs within the ratio becomes redundant and unnecessary. However, in some edge cases we can't have _static validation_ for the code to not have negative numbers in case T is signed.
That's unfortunate, but removing support for signed T feels like an overkill, which would only tackle hypothetical problems where developers control the object construction manually (thus `new + new_neg`, `new_raw + new_raw_signed` methods allowing the sign control).
Perhaps, we could improve some APIs there to make it transparently safer with negative numbers. 
- `unsigned T` is considered to be the _main_ scenario and this is not a problem
- `signed T` is only supported for flexibility, but also safe if users control sign on construction (choosing correct new/new_neg method)
- Perhaps, we could implement extra logic in the constructor that would add sign safety for signed types only (and eliminated on compilation with unsigned T). That feels like a good Hacktoberfest project :)

File: src/fraction/approx/sqrt.rs [hunk 9]
```diff
@@ -0,0 +1,637 @@
...
+    }
+
+    /// Returns a rational approximation of the principal square root of `self`. `decimal_places`
+    /// refers to the accuracy of the square of the return value relative to `self`.
+    ///
+    /// See the [module-level documentation](`self`) for more details.
+    ///
+    /// # Panics
+    /// This method will panic if `self` is negative.
+    pub fn sqrt(&self, decimal_places: u32) -> GenericFraction<BigUint> {
```
Comment (dnsl48) [inline src/fraction/approx/sqrt.rs:510]: I feel like `panic` is an overkill here.

Imagining this code used in external applications where people would use libraries using other libraries. The end application might inherently invoke this code with negative user input, which would lead to the entire application going down. Also, given a lot of anecdotes about _security vulnerabilities_ in the applications caused by people being lazy to use longer method names (thus, leaning towards `sqrt` instead of `sqrt_abs`, which also feels like the _default_ method).

I reckon it might be safer to replicate the default Rust `f64` behaviour (and adhere to the Float trait contract) - https://docs.rs/num/latest/num/trait.Float.html#tymethod.sqrt

That would mean, we could return `NaN` instead of triggering the panic in here.

What do you think?

**Review by dnsl48 — CHANGES_REQUESTED** [review]
Thank you, that looks great!

I only have one last concern about the `panic` to discuss, but otherwise is really happy with everything.

File: src/fraction/approx/sqrt.rs (hunk 9 above)
Comment (squ1dd13) [reply src/fraction/approx/sqrt.rs:510]: Good point here – I hadn't considered comparing to the `f64` API. I guess there's no reason not to return `NaN`, since `GenericFraction` has the ability to represent it. I'll get that fixed. Cheers.

**Review by dnsl48 — CHANGES_REQUESTED** [review]
That looks great, thank you for the contribution @squ1dd13!

That PR looks awesome and I would love to merge it. However, there is a small issue left with "test_raw" taking too long to finish: https://github.com/dnsl48/fraction/actions/runs/6291510982/job/17081838784?pr=84#step:4:660

Looks like it takes over 250s to finish on GitHub (and over 300s on my localhost).

Perhaps we could reduce that time? Ideally the whole `approx` module test suite would run within 30s if possible?
If we want to keep the _heavy_ tests for more thorough coverage, perhaps we should move them into a separate feature (e.g. how it is done with src/tests/division/divide_to_string_u8.rs).

**Comment by squ1dd13** [comment]
That running time had me confused for a minute :D I forgot to say at any point that the algorithm is only remotely "fast" when compiler optimisations are turned on, so I've been testing with `--release`. Indeed it takes around 220s to run `cargo test fraction::approx --features="with-approx"` on my machine; compare this to ~12s to run the same with `--release` enabled[^1].

[^1]: Out of interest, I also tried testing with `RUSTFLAGS="-C target-cpu=native"`, unrealistic as that may be. It reduces the time further down to just under 9s, which I think is just incredible. We're calculating five different square roots to 10, 100, 1k, 10k and 100k decimal places, using numbers represented as `Vec`s, and it's capable of executing in that time. `rustc` is a fantastic thing...

This difference is absolutely not ideal, because it's likely to catch users out. [RFC 2412](https://rust-lang.github.io/rfcs/2412-optimize-attr.html) should help with this, because we would be able to force speed optimisation for specifically the `sqrt` module when the user doesn't supply any other optimisation parameters. For now, I guess the only solution for this aspect of the problem is to add a note to the documentation.

In terms of the testing time, would we be able to test with optimisations enabled? All we'd need is this:

```toml
[profile.test]
opt-level = 3
```

Alternatively, we could test using `--release`, but that seems unnecessary and also has the issue of disabling debug assertions etc.

For now I'll add a note to the documentation and change the build config to optimise tests, but I'm open to reversing those changes and discussing other options.

**Review by dnsl48 — CHANGES_REQUESTED** [review]
Thank you @squ1dd13,
the PR looks great and the tests performance is okay now (happy to go with `opt-level=3` for now).

It looks like we only need to fix benchmarks and it is good to merge

```rust
error[E0599]: no method named `sqrt_raw` found for enum `GenericFraction` in the current scope
  --> benches/bench_fraction.rs:69:28
   |
69 |             b.iter(|| num2.sqrt_raw(n));
   |                            ^^^^^^^^ method not found in `GenericFraction<u8>`

error[E0599]: no method named `sqrt_raw` found for enum `GenericFraction` in the current scope
  --> benches/bench_fraction.rs:73:33
   |
73 |             b.iter(|| small_num.sqrt_raw(n));
   |                                 ^^^^^^^^ method not found in `GenericFraction<BigUint>`

error[E0599]: no method named `sqrt_raw` found for enum `GenericFraction` in the current scope
  --> benches/bench_fraction.rs:77:31
   |
77 |             b.iter(|| big_num.sqrt_raw(n));
   |                               ^^^^^^^^ method not found in `GenericFraction<BigUint>`

For more information about this error, try `rustc --explain E0599`.
error: could not compile `fraction` due to 3 previous errors
warning: build failed, waiting for other jobs to finish...
```

**Comment by squ1dd13** [comment]
Oops, forgot to check that. I've left the benches in but they'll now only run when `with-approx` is enabled.

**Review by dnsl48 — APPROVED** [review]
Thank you for the contribution!

**Comment by dnsl48** [comment]
Released with `0.14.0`

**Comment by squ1dd13** [comment]
Great, thanks!
